assets/js/components/DashboardMainApp.js

```diff
@@ -61,21 +61,56 @@ import { CORE_USER } from '../googlesitekit/datastore/user/constants';
 import { CORE_WIDGETS } from '../googlesitekit/widgets/datastore/constants';
 import { useFeature } from '../hooks/useFeature';
 import useViewOnly from '../hooks/useViewOnly';
-const { useSelect } = Data;
+import { CORE_FORMS } from '../googlesitekit/datastore/forms/constants';
+import { CORE_SITE } from '../googlesitekit/datastore/site/constants';
+import {
+	FORM_CUSTOM_DIMENSIONS_CREATE,
+	MODULES_ANALYTICS_4,
+} from '../modules/analytics-4/datastore/constants';
+const { useSelect, useDispatch } = Data;
 
 export default function DashboardMainApp() {
 	const [ showSurveyPortal, setShowSurveyPortal ] = useState( false );
 
 	const userInputEnabled = useFeature( 'userInput' );
+	const newsKeyMetricsEnabled = useFeature( 'newsKeyMetrics' );
 	const viewOnlyDashboard = useViewOnly();
 
+	const isKeyMetricsSetupCompleted = useSelect( ( select ) =>
+		select( CORE_SITE ).isKeyMetricsSetupCompleted()
+	);
+
+	const autoSubmit = useSelect( ( select ) =>
+		select( CORE_FORMS ).getValue(
+			FORM_CUSTOM_DIMENSIONS_CREATE,
+			'autoSubmit'
+		)
+	);
+
+	const { createCustomDimensions } = useDispatch( MODULES_ANALYTICS_4 );
+
 	useMount( () => {
 		if ( ! viewOnlyDashboard ) {
 			// Render the current survey portal in 5 seconds after the initial rendering.
 			setTimeout( () => setShowSurveyPortal( true ), 5000 );
 		}
 	} );
 
+	useEffect( () => {
+		if (
+			newsKeyMetricsEnabled &&
+			isKeyMetricsSetupCompleted &&
+			autoSubmit
+		) {
+			createCustomDimensions().catch( () => {} );
+		}
```

Comment:
I hope you folks do not mind me chiming in here.

One additional condition to consider here is to make sure the user has the Analytics "edit" scope at this point so that if in case the OAuth flow fails/is cancelled for any reason, we don't attempt to create the custom dimensions.

One other point is that we probably should consider changing the `autoSubmit` flag to `false` once the custom dimensions have been created so that this is not attempted again after they've been already created.

Reply:
Nahid, thank you for the added observations. Good point about the Scope check, I didn't consider the cancel flow there!

As for setting the `autoSubmit` flag back to `false`, Should we do it unconditionally there, or only do it if the action is successful? 

Cheers.

Reply:
Thank you @kuasha420.

> As for setting the `autoSubmit` flag back to `false`, Should we do it unconditionally there, or only do it if the action is successful?

I think we would want to do it only when the action was successfully completed. I mean, certain `createCustomDimension` fetch calls may have encountered errors in the process, but as long as the `createCustomDimensions` action was completed successfully (makes fetch request for all missing custom dimensions & syncs the available custom dimensions), we should set `autoSubmit` flag back to `false`.

Alternatively or additionally, besides existing conditions, we may also want to dispatch this action here only if there are missing custom dimensions.

Let me know what you think, thanks!

Reply:
The "lifetime" of the `autoSubmit` is very fragile, as it gets wiped on refresh. I don't think we need to worry too much about it, as we're not handling error here and the errors will be surfaced in the Tiles themselves if there's a missing custom dimension (as far as I recall). 

Also, the `createCustomDimension` only hits the API if there's missing dimensions, there's a check in the action itself. 

Cheers.

Reply:
> Also, the `createCustomDimension` only hits the API if there's missing dimensions, there's a check in the action itself.

Ah, you're right. Sounds good, thanks!